core/src/Cuda/Kokkos_Cuda_Parallel.hpp

```diff
@@ -531,6 +509,60 @@ class TeamPolicyInternal<Kokkos::Cuda, Properties...>
     return *this;
   }
 #endif
+
+  template <class FunctorType, class ClosureType>
+  int internal_team_size_max(const FunctorType& f) const {
+    using closure_type = ClosureType;
+    typedef Impl::FunctorValueTraits<FunctorType, typename traits::work_tag>
+        functor_value_traits;
+
+    cudaFuncAttributes attr =
+        CudaParallelLaunch<closure_type, typename traits::launch_bounds>::
+            get_cuda_func_attributes();
+    const int block_size =
+        Kokkos::Impl::cuda_get_max_block_size<FunctorType,
+                                              typename traits::launch_bounds>(
+            space().impl_internal_space_instance(), attr, f,
+            (size_t)vector_length(),
+            (size_t)team_scratch_size(0) + 2 * sizeof(double),
+            (size_t)thread_scratch_size(0) + sizeof(double) +
+                ((functor_value_traits::StaticValueSize != 0)
+                     ? 0
+                     : functor_value_traits::value_size(f)));
+
+    // Currently we require Power-of-2 team size for reductions.
```

Comment:
Please assert that `block_size > 0`, since that's an implicit assumption in the code below and the reader may not understand that code/think it's buggy if they don't know that